components/api_server/src/model/actions.py

```diff
@@ -7,73 +7,66 @@
 from shared.model.metric import Metric
 from shared.model.source import Source
 from shared.utils.type import ItemId
-from shared_data_model import DATA_MODEL
 
 from model.report import Report
 from utils.functions import uuid
 from utils.type import Position
 
 
-def copy_item(item, **kwargs):
+def copy_item[Item: Metric | Report | Source | Subject](item: Item, **kwargs) -> Item:
     """Return a copy of the item."""
-    item_copy = item.copy()
+    item_copy = cast(Item, item.copy())
     for key, value in kwargs.items():
         item_copy[key] = value
     return item_copy
 
 
-def copy_source(source, change_name: bool = True):
+def copy_source(source: Source) -> Source:
     """Return a copy of the source."""
-    kwargs = {}
-    if change_name:
-        kwargs["name"] = f"{source.get('name') or DATA_MODEL.sources[source['type']].name} (copy)"
-    return copy_item(source, **kwargs)
+    return copy_item(source)
 
 
-def copy_metric(metric, change_name: bool = True):
+def copy_metric(metric: Metric) -> Metric:
     """Return a copy of the metric and its sources."""
     kwargs: dict[str, Any] = {
-        "sources": {uuid(): copy_source(source, change_name=False) for source in metric["sources"].values()},
+        "sources": {uuid(): copy_source(source) for source in metric["sources"].values()},
     }
-    if change_name:
-        kwargs["name"] = f"{metric.get('name') or DATA_MODEL.metrics[metric['type']].name} (copy)"
     return copy_item(metric, **kwargs)
 
 
-def copy_subject(subject, change_name: bool = True):
+def copy_subject(subject: Subject) -> Subject:
     """Return a copy of the subject, its metrics, and their sources."""
     kwargs: dict[str, Any] = {
-        "metrics": {uuid(): copy_metric(metric, change_name=False) for metric in subject["metrics"].values()},
+        "metrics": {uuid(): copy_metric(metric) for metric in subject["metrics"].values()},
     }
-    if change_name:
-        kwargs["name"] = f"{subject.get('name') or DATA_MODEL.all_subjects[subject['type']].name} (copy)"
     return copy_item(subject, **kwargs)
 
 
-def copy_report(report):
+def copy_report(report: Report) -> Report:
     """Return a copy of the report, its subjects, their metrics, and their sources."""
     return copy_item(
         report,
         report_uuid=uuid(),
-        title=f"{report['title']} (copy)",
-        subjects={uuid(): copy_subject(subject, change_name=False) for subject in report["subjects"].values()},
+        subjects={uuid(): copy_subject(subject) for subject in report["subjects"].values()},
     )
 
 
+type ItemsDictType = MutableMapping[ItemId, Metric | Source | Subject]
+
+
 def move_item(
     container: Report | Subject | Metric,
     item_to_move: Subject | Metric | Source,
     new_position: Position,
 ) -> tuple[int, int]:
     """Change the item position."""
-    items_dict_type = MutableMapping[ItemId, Metric | Source | Subject]
-    items_dict: items_dict_type
+    items_dict: ItemsDictType
```

Comment:
Is this magic not something we can implement in a `move_child` implementation of the various classes? It seems like we need to do a lot of housekeeping with these type checks

Reply:
I think these changes should be done as part of refactoring `Metric`, `Source`, etc. to proper domain classes. Also see previous comment.